src/browser/file-recovery-service.js

```diff
@@ -0,0 +1,65 @@
+'use babel'
+
+import {ipcMain} from 'electron'
+import crypto from 'crypto'
+import Path from 'path'
+import fs from 'fs-plus'
+
+export default class FileRecoveryService {
+  constructor (recoveryDirectory) {
+    this.recoveryDirectory = recoveryDirectory
+    this.recoveryPathsByWindowAndFilePath = new WeakMap()
+    this.crashListeners = new WeakSet()
+  }
+
+  start () {
+    ipcMain.on('will-save-path', this.willSavePath.bind(this))
+    ipcMain.on('did-save-path', this.didSavePath.bind(this))
+  }
+
+  willSavePath (event, path) {
+    if (!fs.existsSync(path)) {
+      // Unexisting files won't be truncated/overwritten, and so there's no data to be lost.
+      return
+    }
+
+    const window = event.sender
+    const recoveryFileName = crypto.randomBytes(5).toString('hex')
+    const recoveryPath = Path.join(this.recoveryDirectory, recoveryFileName)
+    fs.writeFileSync(recoveryPath, fs.readFileSync(path))
+
+    if (!this.recoveryPathsByWindowAndFilePath.has(window)) {
+      this.recoveryPathsByWindowAndFilePath.set(window, new Map())
+    }
+    this.recoveryPathsByWindowAndFilePath.get(window).set(path, recoveryPath)
+
+    if (!this.crashListeners.has(window)) {
+      window.on('crashed', () => this.recoverFilesForWindow(window))
```

Comment:
Do you think we should remove the listener when the window closes? Would the callback's reference to `window` retain it and keep it from being removed from the WeakSet?